In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\shyro\OneDrive\Documents\Instacart Basket Analysis'

In [10]:
# import orders_products_merged.pkl as a dataframe
ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data','orders_products_merged.pkl'))

In [19]:
#create a subset
df = ords_prods_merged [:1000000]

In [24]:
df.shape

(1000000, 16)

In [25]:
# define a function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [23]:
# apply the function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\shyro\AppData\Local\Temp\ipykernel_21396\3603513338.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [26]:
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [27]:
df['prices'].max()

14.8

In [30]:
# using loc instead to create a user-defined function
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [29]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [31]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [32]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [34]:
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_range  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

     price_range_loc  
0  Mid-range product  
1  Mid-range product  
2  Mid-range product  
3  Mid-range product  
4  Mid-range product

In [35]:
#create a user-defined function using loc for the entire dataframe ords_prods_merged
[ords_prods]_merged.loc[ords_prods_merged['prices']> 15,'price_range_loc'] = 'High-range product'

In [37]:
ords_prods_merged.loc[(ords_prods_merged['prices']<= 15) & (ords_prods_merged['prices']> 5),'price_range_loc'] = 'Mid-range product'

In [38]:
ords_prods_merged.loc[ords_prods_merged['prices']<= 5,'price_range_loc'] = 'Low-range product'

In [39]:
ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

In [42]:
# frequency of the "order_dow" column
ords_prods_merged['order_dow'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_dow, dtype: int64

In [44]:
# create a for-loop for order_dow

result = []

for value in ords_prods_merged["order_dow"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [45]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [46]:
# merge for-loops with ords_prods_merged dataframe
ords_prods_merged['busiest_day'] = result

In [47]:
ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

In [48]:
ords_prods_merged.shape

(32404859, 17)

In [50]:
ords_prods_merged.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

## Task 4.7

### 2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [53]:
# get frequency of orders_day_of_week
ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [54]:
# create a for-loop for the 2 busiest and slowest days

result_2 = []

for value in ords_prods_merged["orders_day_of_week"]:
  if value == 0 or value ==1:
    result_2.append("Busiest day")
  elif value == 4 or value == 3:
    result_2.append("Least busy")
  else:
    result_2.append("Regularly busy")

In [55]:
# merge for-loops with ords_prods_merged dataframe
ords_prods_merged['busiest_days'] = result_2

In [56]:
ords_prods_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

     price_range_loc     busiest_day    busiest_days  
0  Mid-range product  Regularly busy  Regularly busy  
1  Mid-range product  Regularly busy      Least busy  
2  Mid-range product  Regularly busy      Least busy  
3  Mid-range product      Least busy      Least busy  
4  Mid-range product      Least busy      Least busy

### 3. Check the values of this new column for accuracy. Note any observations in markdown format.


In [57]:
ords_prods_merged.shape

(32404859, 18)

#### The dataframe has the same amount of rows (32404859) but the columns increased by 1, from 17 to 18

### 4. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [58]:
# get frequency of order_hour_of_day
ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

#### Observations: there are 24 hours in a day divided by 3 categories (Most orders, Average orders, and Fewest orders). A block of 8 8 hours will be used in the for-loop condition. 

In [59]:
# create a for-loop for the busiest hours of the day

hours = []

for value in ords_prods_merged["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12 ,16, 9]:
    hours.append("Most orders")
  elif value in [17, 8, 18, 19, 20, 7, 21, 22]:
    hours.append("Average orders")
  else:
    hours.append("Fewest orders")

In [60]:
# merge busiest hours of the day for-loops with ords_prods_merged dataframe
ords_prods_merged['busiest_hours_of_day'] = hours

In [62]:
#check the output for order_hour_of_day and busiest_hours_of_day
ords_prods_merged[['order_hour_of_day', 'busiest_hours_of_day']].head()

order_hour_of_day busiest_hours_of_day
0                  8       Average orders
1                  7       Average orders
2                 12          Most orders
3                  7       Average orders
4                 15          Most orders

### 5. Print the frequency for this new column.

In [63]:
ords_prods_merged['busiest_hours_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_hours_of_day, dtype: int64

### 8. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.


In [64]:
ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data','ords_prods_merged.pkl'))